lib/util/Components.js

```diff
@@ -247,6 +247,40 @@ function getWrapperFunctions(context, pragma) {
   ]);
 }
 
+// eslint-disable-next-line valid-jsdoc
+/**
+ * Merge many eslint rules into one
+ * @param {{[_: string]: Function}[]} rules the returned values for eslint rule.create(context)
+ * @returns {{[_: string]: Function}} merged rule
+ */
+function mergeRules(rules) {
+  /** @type {Map<string, Function[]>} */
+  const handlersByKey = new Map();
+  for (const rule of rules) {
+    for (const key of Object.keys(rule)) {
+      const fns = handlersByKey.get(key);
+      if (!fns) {
+        handlersByKey.set(key, [rule[key]]);
+      } else {
+        fns.push(rule[key]);
+      }
+    }
+  }
+
+  /** @type {{[key: string]: Function}} */
+  const rule = {};
+  for (const key of handlersByKey.keys()) {
+    const fns = handlersByKey.get(key);
+    rule[key] = function mergedHandler(node) {
+      for (const fn of fns) {
```

Comment:
for-of is slow; `fns.forEach((node) => fn(node))` is almost certainly faster (and cleaner; loops are bad)

Reply:
for-of is not slow in node, v8 have done optimizations to make it performs comparable to for loop.

```js
const N = 100000;
const length = 100;

const array = new Array(length).fill().map((_, i) => i);

function benchmarkForLoop() {
  let start = performance.now();
  let x = 0;
  for (let i = 0; i < N; i++) {
    for (let j = 0; j < array.length; j++) {
      x = (x + array[j]) % 7;
    }
  }
  let end = performance.now();
  console.log(end - start);
}

function benchmarkForOfLoop() {
  let start = performance.now();
  let x = 0;
  for (let i = 0; i < N; i++) {
    for (const y of array) {
      x = (x + y) % 7;
    }
  }
  let end = performance.now();
  console.log(end - start);
}

function benchmarkForEach() {
  let start = performance.now();
  let x = 0;
  for (let i = 0; i < N; i++) {
    array.forEach((y) => {
      x = (x + y) % 7;
    });
  }
  let end = performance.now();
  console.log(end - start);
}

benchmarkForLoop();
benchmarkForOfLoop();
benchmarkForEach();

// Output:
// 42.265081003308296
// 47.21948900818825
// 57.372923001646996
```

Reply:
Try those benchmarks without using Array(), which creates a sparse array that .fill may not be able to un-taint.